modules/ztp-nbde-rekeying-all-nbde-nodes.adoc

```diff
@@ -0,0 +1,124 @@
+.Procedure
+
+You can rekey all of the nodes on a remote cluster using a daemonset object without incurring any downtime to the remote cluster.
+
+. Create a daemonset object based on the following template. This template is for a setup with three redundant Tang servers, but can be easily adapted to other situations. Change the Tang server URLs and thumbprints in the `NEW_TANG_PIN` environment to suit your environment:
++
+[source,yaml]
+----
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: tang-rekey
+  namespace: openshift-machine-config-operator
+spec:
+  selector:
+    matchLabels:
+      name: tang-rekey
+  template:
+    metadata:
+      labels:
+        name: tang-rekey
+    spec:
+      containers:
+      - name: tang-rekey
+        image: registry.access.redhat.com/ubi8/ubi-minimal:8.4
+        imagePullPolicy: IfNotPresent
+        command:
+        - "/sbin/chroot"
+        - "/host"
+        - "/bin/bash"
+        - "-ec"
+        args:
+        - |
+          rm -f /tmp/rekey-complete || true
+          echo "Current tang pin:"
+          clevis-luks-list -d $ROOT_DEV -s 1
+          echo "Applying new tang pin: $NEW_TANG_PIN"
+          clevis-luks-edit -f -d $ROOT_DEV -s 1 -c "$NEW_TANG_PIN"
+          echo "Pin applied successfully"
+          touch /tmp/rekey-complete
+          sleep infinity
+        readinessProbe:
+          exec:
+            command:
+            - cat
+            - /host/tmp/rekey-complete
+          initialDelaySeconds: 30
+          periodSeconds: 10
+        env:
+        - name: ROOT_DEV
+          value: /dev/sda4
+        - name: NEW_TANG_PIN
+          value: >-
+            {"t":1,"pins":{"tang":[
+              {"url":"http://tangserver01:7500","thp":"WOjQYkyK7DxY_T5pMncMO5w0f6E"},
+              {"url":"http://tangserver02:7500","thp":"I5Ynh2JefoAO3tNH9TgI4obIaXI"},
+              {"url":"http://tangserver03:7500","thp":"38qWZVeDKzCPG9pHLqKzs6k1ons"}
+            ]}}
+        volumeMounts:
+        - name: hostroot
+          mountPath: /host
+        securityContext:
+          privileged: true
+      volumes:
+      - name: hostroot
+        hostPath:
+          path: /
+      nodeSelector:
+        kubernetes.io/os: linux
+      priorityClassName: system-node-critical
+      restartPolicy: Always
+      serviceAccount: machine-config-daemon
+      serviceAccountName: machine-config-daemon
+----
++
+In this case, even though you are rekeying `tangserver01`, you must specify not only the new thumbprint for `tangserver01`, but also the current thumbprints for all other Tang servers.  Failure to specify all thumbprints for a rekeying operation opens up the opportunity for a man-in-the-middle attack.
+
+. Distribute the daemonset to every cluster that needs to be rekeyed. This can be as simple as executing `oc apply -f tang-rekey.yaml` for a single cluster, but to execute at scale, wrap the daemonset in an ACM policy, with a placement rule to apply it to the appropriate set of clusters.
```

Comment:
After discussing this with the ACM people, I'd like to change this paragraph to the following:

```
Distribute the daemonset to every cluster that needs to be rekeyed. This can be as simple as executing `oc apply -f tang-rekey.yaml` for a single cluster, but to execute at scale, wrap the daemonset in an ACM policy. This ACM configuration should contain one policy to deploy the daemonset, a second policy to check that all the daemonset pods are READY, and a placement rule to apply it to the appropriate set of clusters.
```

Note: Needs a pass from y'all for grammar / style / de-engineerification

Reply:
change made.